csharp/src/Ice/OutputStream.cs

```diff
@@ -1247,7 +1307,7 @@ private void Expand(int n)
                 size = Math.Max(n - remaining, size);
                 byte[] buffer = new byte[size];
                 _segmentList.Add(buffer);
-                if (_segmentList.Count == 1)
+                if (_segmentList.Count == 1) // TODO: missing comment. This occurs only when _currentSegment is empty?
                 {
                     _currentSegment = buffer;
                 }
```

Comment:
```suggestion
                // If the current segment is empty or if there isn't any remaining bytes
                // update the current segment to the newly created segment
                if (_segmentList.Count == 1 || remaining == 0)
                {
                    _currentSegment = buffer;
                }
```

Reply:
Seems it is not working as I will expect getting a failure after this change

Reply:
Is this `|| remaining == 0` correct?

It seems to break the guarantee that `_tail.Segment` always points to `_currentSegment`.

Reply:
You are write the fix should be  more like 

```
                if (_segmentList.Count == 1)
                {
                    _currentSegment = buffer;
                }
                else if (remaining == 0)
                {
                    _currentSegment = buffer;
                    _tail.Segment++;
                    _tail.Offset = 0;
                }
```

Reply:
Ok, makes sense.

Reply:
Fixed as suggested.